readme.md
# Hi, there 👋

`search-sdk` 是一款为 spring 开发者集成的插件，您可以使用它来从指定的搜索引擎中获取结果。

**v0.0.1 版本功能比较单一，只能获取指定搜索引擎的至多 10 条搜索结果。**

## ✨ 特征

- [x] 基于 Playwright 模拟从 Brave、Google、Bing 获取搜索结果
- [ ] 支持自定义获取搜索结果的数量、分页、导出
- [ ] 集成 Bing API、Google API
- [ ] 支持 自定义 Brave Goggles 搜索权重
- [ ] 支持 DuckDuckgo 等更多的搜索引擎

## 🎉 使用

### 步骤 1: 增加依赖

```pom.xml
<dependency>
    <groupId>io.github.unickcheng</groupId>
    <artifactId>search-sdk</artifactId>
    <version>0.0.1</version>
</dependency>
```
> 注意当前还未提交到 Maven 中央仓库，因此您需要额外地添加下述镜像地址在您的 pom.xml 中
```
<repositories>
    <repository>
        <id>unickcheng</id>
        <url>https://unickcheng.github.io/maven-packages/</url>
    </repository>
</repositories>
```

### 步骤2：创建 Bean

```java
@Bean
public SearchEngineService searchEngineService () {
	return new SearchEngineService();
}
```

>? 如果您不知道如何开始使用，您可以参考 [spring-boot-demo](https://github.com/UNICKCHENG/search-sdk/tree/main/spring-boot-demo)

### 如果您是前端人员

您可以直接使用 [Playwright](https://www.npmjs.com/package/playwright) 官方提供的依赖，当然，您也可使用 [spring-boot-demo](https://github.com/UNICKCHENG/search-sdk/tree/main/spring-boot-demo) 作为现成的后端接口来使用。

在启动后端项目之前，您需要确保已经配置好 `java 11` 和 `Maven 3.0+`。然后您需要到 spring-boot-demo 的目录下执行下述命令来运行项目。当然您也可以使用内存大户 IDEA。
```bash
# 下载所需依赖
mvn install
# 启动 spring boot 项目
mvn spring-boot:run
```


## 为什么要创建这个项目

去年 Brave 推出了 [Brave Goggles](https://search.brave.com/help/goggles)，这使得用户可以根据自己的需求来自定义搜索权重。打个比方，如果您是一名学术工作者，您可以为 Wikipedia、Google Scholar、connectedpapers 等网站增加相应的权重，这样将使得和这些网站相关的搜索结果更加靠前。

于是，我开始尝试能不能将它集成到 wechat 中去，像 telegram、discord 那样通过斜杆命令，直接获取相应的搜索结果。遗憾的是，Brave 官方似乎并没有像 Google 和 Bing 一样提供 API，虽然有 [SerpApi](https://serpapi.com/) 这类第三方接口，但是对于个人开发者来说并不太友善。

既然如此，那只能造个轮子呗~

## 💖 感谢

- https://playwright.dev/
- [Spring Boot](https://spring.io/projects/spring-boot) provides framework support
- [Open Source Dependencies](https://github.com/UNICKCHENG/search-sdk/network/dependencies) provides support
- Thanks to all open source projects for sharing ideas and techniques


CHANGELOG.md
## 0.0.1 (2023-02-14)


### Features

* **search-sdk:** initialize basic functions ([30f0bfb](https://github.com/UNICKCHENG/search-sdk/commit/30f0bfb0c3dbe8da1dce7463f2c6b9b6a490bd13))




src/main/java/io/github/unickcheng/searchsdk/SearchResult.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package io.github.unickcheng.searchsdk;

import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

/**
 * @author unickcheng
 * @since 2023-02-14
 */

@Getter
@Setter
@Builder
public class SearchResult {
    Integer id;
    String url;
    String title;
}


src/main/java/io/github/unickcheng/searchsdk/SearchEngineEnum.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package io.github.unickcheng.searchsdk;

import lombok.AllArgsConstructor;
import lombok.Getter;

/**
 * Built-in configuration for common search engines
 * @author unickcheng
 * @since 2023-02-14
 */

@Getter
@AllArgsConstructor
public enum SearchEngineEnum {

    GOOGLE("https://www.google.com/search?q=%s"
            , "#search #rso > div"
            , "h3"
            ,"a[href^='http']"),
    BING("https://www.bing.com/search?q=%s"
            , "#b_results .b_algo"
            , "h2 > a"
            , "h2 > a"),
    BRAVE("https://search.brave.com/search?q=%s"
            , "#results > div"
            , "a > span"
            , "a");

    /**
     * 访问链接
     * <p>example : <a href="https://www.google.com">https://www.google.com</a>
     */
    private final String visitParameters;

    /**
     * 获取全部结果的 css 定位条件
     * <p>example: .b_algo
     * <p>ref: <a href="https://playwright.dev/java/docs/api/class-elementhandle#element-handle-query-selector-all">handle-query-selector-all</a>
     */
    private final String selectorAllParameters;

    /**
     * 获取标题的 css 定位条件
     * <p>example: h2 > a
     * <p>ref: <a href="https://playwright.dev/java/docs/api/class-elementhandle#element-handle-query-selector">handle-query-selector</a>
     */
    private final String selectorTitleParameters;

    /**
     * 获取链接的 css 定位条件
     * <p>example: h2 > a
     * <p>ref <a href="https://playwright.dev/java/docs/api/class-elementhandle#element-handle-query-selector">handle-query-selector</a>
     */
    private final String selectorHrefParameters;
}


src/main/java/io/github/unickcheng/searchsdk/SearchEngineService.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package io.github.unickcheng.searchsdk;

import com.microsoft.playwright.*;
import com.microsoft.playwright.options.Proxy;
import io.github.unickcheng.rhandler.exception.RHandlerException;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

import javax.annotation.PostConstruct;
import java.util.ArrayList;
import java.util.List;

/**
 * @author unickcheng
 * @since 2023-02-14
 */
@Slf4j
@Getter
@SuppressWarnings("unused")
public class SearchEngineService {

    private final Playwright playwright;
    private final Browser browser;

    public SearchEngineService() {
        try {
            this.playwright = Playwright.create();
            this.browser = playwright.chromium().launch(
                new BrowserType.LaunchOptions()
            );
        } catch (Exception e) {
            throw new RHandlerException("Failed initializing SearchEngineService");
        }
    }
    public SearchEngineService (Proxy proxy) {
        try {
            this.playwright = Playwright.create();
            this.browser = playwright.chromium().launch(
                    new BrowserType.LaunchOptions().setProxy(proxy)
            );
        } catch (Exception e) {
            throw new RHandlerException("Failed initializing SearchEngineService");
        }
    }
    public SearchEngineService (Playwright playwright, Browser browser) {
        this.playwright = playwright;
        this.browser = browser;
    }

    @PostConstruct
    public void postConstruct () {
        log.info("Hi, there! SearchEngineService has been initialized successfully.");
    }

    /**
     * Specify a search engine for your query <br/>
     * 使用 playwright 模拟页面操作获取查询结果, 并将结果集进行处理返回 <br/>
     * TODO 当前最多返回 10 条数据集, 且不支持限制数据操作
     * @param searchEngineEnum 搜索引擎
     * @param content 搜索文本
     * @return 查询结果列表
     */
    public List<SearchResult> search (SearchEngineEnum searchEngineEnum, String content) {
        try {
            List<SearchResult> data = new ArrayList<>();
            Page page = this.browser.newPage();

            page.navigate(
                    String.format(
                            searchEngineEnum.getVisitParameters(),
                            content.trim().replaceAll(" ", "+")
                    )
            );
            List<ElementHandle> re = page.querySelectorAll(searchEngineEnum.getSelectorAllParameters());
            for (int i = 0, cnt = 0, len = re.size(); i < len && cnt < 10; ++i) {
                try {
                    String title = re.get(i).querySelector(searchEngineEnum.getSelectorTitleParameters()).textContent().replaceAll("\\n", "").trim();
                    String url = re.get(i).querySelector(searchEngineEnum.getSelectorHrefParameters()).getAttribute("href");
                    data.add(SearchResult.builder().id(++cnt).title(title).url(url).build());
                } catch (Exception ignored) {
                }
            }
            page.close();
            return data;
        } catch (Exception ex) {
            throw new RHandlerException("Failed to search.");
        }
    }
}



spring-boot-demo/src/main/java/cc/unickcheng/demo/SearchEngineConfig.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package cc.unickcheng.demo;

import io.github.unickcheng.searchsdk.SearchEngineService;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * @author unickcheng
 * @since 2023-02-14
 */

@Configuration
public class SearchEngineConfig {

    // 默认方式
    @Bean
    public SearchEngineService searchEngineService () {
        return new SearchEngineService();
    }

    // 自定义代理方式
//    @Bean
//    public SearchEngineService searchEngineService () {
//        return new SearchEngineService(new Proxy("http://127.0.0.1:9999"));
//    }
//
    // 自定义初始化方式
//    @Bean
//    public SearchEngineService searchEngineService () {
//        try {
//            Playwright playwright = Playwright.create();
//            Browser browser = playwright.chromium().launch(
//                    new BrowserType.LaunchOptions()
//            );
//            return new SearchEngineService(playwright, browser);
//        } catch (Exception e) {
//            throw new RHandlerException("Failed initializing SearchEngineService");
//        }
//    }
}


spring-boot-demo/src/main/java/cc/unickcheng/demo/SearchEngineApplication.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package cc.unickcheng.demo;

import lombok.extern.slf4j.Slf4j;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

/**
 * After running the demo, the web side visits localhost:8080/openapi-ui.html
 * @author unickcheng
 * @since 2023-02-14
 */

@Slf4j
@SpringBootApplication
public class SearchEngineApplication {
    public static void main (String[] args) {
        SpringApplication.run(SearchEngineApplication.class, args);
        log.info(">>> {}", SearchEngineApplication.class.getSimpleName().toUpperCase() +
                " STARTING SUCCESS. See https://github.com/UNICKCHENG/search-sdk for more information. ");
    }
}


spring-boot-demo/src/main/java/cc/unickcheng/demo/SearchEngineController.java
/*
 * Copyright 2002-2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package cc.unickcheng.demo;

import io.github.unickcheng.rhandler.annotation.RHandlerController;
import io.github.unickcheng.searchsdk.SearchEngineEnum;
import io.github.unickcheng.searchsdk.SearchEngineService;
import io.github.unickcheng.searchsdk.SearchResult;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.Parameters;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;

import javax.annotation.Resource;

/**
 * After running the demo, the web side visits localhost:8080/openapi-ui.html <br/>
 * demo 运行后, 网页端请访问 localhost:8080/openapi-ui.html
 * @author unickcheng
 * @since 2023-02-14
 */

@RHandlerController
public class SearchEngineController {
    @Resource
    private SearchEngineService searchEngineService;

    @Operation(description = "Google Search Api")
    @Parameters({
            @Parameter(name = "q", description = "Ask me a question...", required = true)
    })
    @GetMapping("google")
    public java.util.List<SearchResult> searchByGoogle (@RequestParam(value = "q") String content) {
        return searchEngineService.search(SearchEngineEnum.GOOGLE, content);
    }

    @Operation(description = "Bing Search Api")
    @Parameters({
            @Parameter(name = "q", description = "Ask me a question...", required = true)
    })
    @GetMapping("bing")
    public java.util.List<SearchResult> searchByBing (@RequestParam(value = "q") String content) {
        return searchEngineService.search(SearchEngineEnum.BING, content);
    }

    @Operation(description = "Brave Search Api")
    @Parameters({
            @Parameter(name = "q", description = "Ask me a question...", required = true)
    })
    @GetMapping("brave")
    public java.util.List<SearchResult> searchByBrave (@RequestParam(value = "q") String content) {
        return searchEngineService.search(SearchEngineEnum.BRAVE, content);
    }
}
